core/src/stored/append.cc

```diff
@@ -52,39 +50,151 @@ static char OK_replicate[] = "3000 OK replicate data\n";
+    , data_len_(record->data_len)
+    , data_(record->data, record->data + record->data_len)
+{
+}
+
+DeviceRecord ProcessedFileData::GetData()
+{
+  DeviceRecord devicerecord{};
+  devicerecord.VolSessionId = volsessionid_;
+  devicerecord.VolSessionTime = volsessiontime_;
+  devicerecord.FileIndex = fileindex_;
+  devicerecord.Stream = stream_;
+  devicerecord.data_len = data_len_;
+  devicerecord.data = data_.data();
+
+  return devicerecord;
+}
+
+ProcessedFile::ProcessedFile(int32_t fileindex) : fileindex_(fileindex) {}
+
+void ProcessedFile::SendAttributesToDirector(JobControlRecord* jcr)
+{
+  for_each(attributes_.begin(), attributes_.end(),
+           [&jcr](ProcessedFileData& attribute) {
+             DeviceRecord devicerecord = attribute.GetData();
+             SendAttrsToDir(jcr, &devicerecord);
+           });
+}
+
+void ProcessedFile::AddAttribute(DeviceRecord* record)
+{
+  attributes_.emplace_back(ProcessedFileData(record));
+}
+
+bool IsAttribute(DeviceRecord* record)
+{
+  return record->maskedStream == STREAM_UNIX_ATTRIBUTES
+         || record->maskedStream == STREAM_UNIX_ATTRIBUTES_EX
+         || record->maskedStream == STREAM_RESTORE_OBJECT
+         || CryptoDigestStreamType(record->maskedStream) != CRYPTO_DIGEST_NONE;
+}
+
+static void UpdateFileList(JobControlRecord* jcr)
+{
+  Dmsg0(100, _("... update file list\n"));
+  jcr->impl->dcr->DirAskToUpdateFileList(jcr);
+}
+
+static void UpdateJobmediaRecord(JobControlRecord* jcr)
+{
+  Dmsg0(100, _("... create job media record\n"));
+  jcr->impl->dcr->DirCreateJobmediaRecord(false);
+}
+
+static void UpdateJobrecord(JobControlRecord* jcr)
+{
+  Dmsg2(100, _("... update job record: %llu bytes %lu files\n"), jcr->JobBytes,
+        jcr->JobFiles);
+  jcr->impl->dcr->DirAskToUpdateJobRecord(jcr);
+}
+
+void DoBackupCheckpoint(JobControlRecord* jcr)
+{
+  Jmsg(jcr, M_INFO, 0,
+       _("Checkpoint: Syncing current backup status to catalog\n"));
+  UpdateJobrecord(jcr);
+  UpdateFileList(jcr);
+  UpdateJobmediaRecord(jcr);
+  Jmsg(jcr, M_INFO, 0, _("Checkpoint completed\n"));
+}
+
+static time_t DoTimedCheckpoint(JobControlRecord* jcr,
+                                time_t checkpoint_time,
+                                time_t checkpoint_interval)
+{
+  time_t now
+      = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+  if (now > checkpoint_time) {
+    checkpoint_time = now + checkpoint_interval;
+    Jmsg(jcr, M_INFO, 0,
+         _("Doing timed backup checkpoint. Next checkpoint in %d seconds\n"),
+         checkpoint_interval);
+    DoBackupCheckpoint(jcr);
+  }
+
+  return checkpoint_time;
+}
+
+static void SaveFullyProcessedFiles(JobControlRecord* jcr,
+                                    std::vector<ProcessedFile>& processed_files)
+{
+  if (!processed_files.empty()) {
+    for_each(
+        processed_files.begin(), processed_files.end(),
+        [&jcr](ProcessedFile& file) { file.SendAttributesToDirector(jcr); });
```

Comment:
when `SendAttributesToDirector()` is marked const, you can pass ProcessedFile as const&
```suggestion
    for_each(
        processed_files.begin(), processed_files.end(),
        [&jcr](const ProcessedFile& file) { file.SendAttributesToDirector(jcr); });
```